### PR TITLE
Show FixAll progress when the user invokes a long-runnning light bulb.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
This involved threading an IProgressTracker through all the FixAll code, the code in the analyzer subsystem to compute diagnostics for a project, and the code in the workspace to apply a large solution change to a project.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
image

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide 
Note: this depends on VS2015 Update2.  @davkean @jasonmalinowski tells me that jenkins is till on Update1.  Is there an ETA on when they'd be updated?

**Comment by robinsedlaczek** [comment]
Does it relate to #2627?

**Comment by robinsedlaczek** [comment]
Sorry, #2637 is the issue I refer to.

File: src/Workspaces/Core/Portable/Shared/Utilities/IProgressTracker.cs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+            }
+        }
+
+        public void ItemCompleted()
+        {
+            lock (gate)
+            {
+                _completedItems++;
+            }
+        }
```
Comment (balajikris) [inline src/Workspaces/Core/Portable/Shared/Utilities/IProgressTracker.cs]: just noting that `Interlocked` api can handle all the above operations without a lock. any specific reasons to choose `lock` over lock-free `Interlocked` ?

Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Utilities/IProgressTracker.cs]: Good point!

**Comment by CyrusNajmabadi** [comment]
@robinsedlaczek Yes, this relates to that.  We now at least show a progress bar.  And, for solution-wide fix-all actions, we'll show you the progress as we both compute diagnostics for the solution and the progress as we apply the fixes to every project in the solution.

**Comment by CyrusNajmabadi** [comment]
also @Pilchie Is there a way to know if we're on Update1 or Update2?  I need to not try to update progress if we're on Update1.

**Comment by Pilchie** [comment]
I believe you can test the version number of devenv.exe - we make a point of updating it in every update.

**Comment by balajikris** [comment]
looks like roslyn-internal needs corresponding changes since IWaitIndicator changed.

File: src/EditorFeatures/Core/Host/IWaitIndicator.cs [hunk 2]
```diff
@@ -9,7 +9,16 @@ internal interface IWaitIndicator
         /// <summary>
         /// Schedule the action on the caller's thread and wait for the task to complete.
         /// </summary>
-        WaitIndicatorResult Wait(string title, string message, bool allowCancel, Action<IWaitContext> action);
-        IWaitContext StartWait(string title, string message, bool allowCancel);
+        WaitIndicatorResult Wait(string title, string message, bool allowCancel, bool showProgress, Action<IWaitContext> action);
+        IWaitContext StartWait(string title, string message, bool allowCancel, bool showProgress);
+    }
+
+    internal static class IWaitIndicatorExtensions
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Host/IWaitIndicator.cs]: Why not just update the callers?

File: src/Workspaces/Core/Portable/Shared/Utilities/IProgressTracker.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        int TotalItems { get; }
+
+        void AddItems(int count);
+        void ItemCompleted();
+        void Clear();
+    }
+
+    internal class SimpleProgressTracker : IProgressTracker
+    {
+        private readonly object gate = new object();
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Shared/Utilities/IProgressTracker.cs]: `_gate`

File: src/Workspaces/Core/Portable/Workspaces.csproj [hunk 4]
```diff
@@ -265,6 +265,7 @@
     <InternalsVisibleTo Include="Microsoft.VisualStudio.InteractiveServices" />
     <InternalsVisibleTo Include="Roslyn.VisualStudio.Setup" />
     <InternalsVisibleTo Include="Microsoft.VisualStudio.VisualBasic.Repl" />
+    <InternalsVisibleToTest Include="Microsoft.CodeAnalysis.Editor.UI.Wpf" />
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspaces.csproj]: This really shouldn't be here.

**Comment by jasonmalinowski** [comment]
:+1:

**Pull request merged**